File: Python/stone.py
from bangtal import *
from config import *
from board import *

class Stone(Object):
    def __init__(self, othello, ix, iy, shown = True):
        super().__init__(Config.IMAGE_STONE_BLANK)

        self.locate(othello, Config.BOARD_INIT_X + Config.STONE_SIZE_X * ix, Config.BOARD_INIT_Y + Config.STONE_SIZE_Y * iy)
        if shown: self.show()

        self.othello = othello
        self.ix = ix
        self.iy = iy

    def onMouseAction(self, x, y, action):
        self.othello.onClick(self.ix, self.iy)

    def setState(self, turn, state):
        if state == State.BLACK: self.setImage(Config.IMAGE_STONE_BLACK)
        elif state == State.WHITE: self.setImage(Config.IMAGE_STONE_WHITE)
        elif state == State.POSSIBLE: self.setImage(Config.IMAGE_STONE_BLACK_POSSIBLE if turn == Turn.BLACK else Config.IMAGE_STONE_WHITE_POSSIBLE)
        else: self.setImage(Config.IMAGE_STONE_BLANK)

File: C++/Stone.cpp
#include "Stone.h"
using namespace bangtal;

#include "config.h"
#include "Othello.h"

StonePtr Stone::create(OthelloPtr othello, int ix, int iy, bool shown)
{
	auto stone = StonePtr(new Stone(othello, ix, iy, shown));
	Object::add(stone);

	return stone;
}

Stone::Stone(OthelloPtr scene, int x, int y, bool shown)
	: Object(IMAGE_STONE_BLANK, scene, BOARD_INIT_X + STONE_SIZE_X * x, BOARD_INIT_Y + STONE_SIZE_Y * y, shown),
	  othello(scene), ix(x), iy(y), state(State::NONE)
{
}

void Stone::setState(Turn turn, State s)
{
	state = s;
	switch (state) {
		case State::BLACK: setImage(IMAGE_STONE_BLACK); break;
		case State::WHITE: setImage(IMAGE_STONE_WHITE); break;
		case State::POSSIBLE: turn == Turn::BLACK ? setImage(IMAGE_STONE_BLACK_POSSIBLE) : setImage(IMAGE_STONE_WHITE_POSSIBLE); break;
		default: setImage(IMAGE_STONE_BLANK); break;
	}
}

bool Stone::onMouse(int x, int y, MouseAction action)
{
	if (state == State::POSSIBLE) {
		return othello->onClick(ix, iy);
	}
	return false;
}